### PR TITLE
easyrsa: 3.0.8 -> 3.1.1

### PR DESCRIPTION
**Pull request by numinit**
###### Description of changes

Update easyrsa to version 3. No more patch required.

Also delete easyrsa 2.x as it was replaced with easyrsa 3 in 2013.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/networking/easyrsa/default.nix [hunk 1]
```diff
@@ -1,33 +1,42 @@
-{ lib, stdenv, fetchFromGitHub, openssl, runtimeShell }:
+{ lib, stdenv, fetchFromGitHub, openssl, makeWrapper, runtimeShell }:
 
 let
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/easyrsa/default.nix:3]:
```suggestion
```

File: pkgs/tools/networking/easyrsa/default.nix [hunk 2]
```diff
@@ -1,33 +1,42 @@
...
-  version = "3.0.8";
+  version = "3.1.1";
+  gitRev = "v${version}";
 in stdenv.mkDerivation {
   pname = "easyrsa";
   inherit version;
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/easyrsa/default.nix:8]:
```suggestion
stdenv.mkDerivation rec {
  pname = "easyrsa";
  version = "3.1.1";
```

File: pkgs/tools/networking/easyrsa/default.nix [hunk 3]
```diff
@@ -1,33 +1,42 @@
...
 
   src = fetchFromGitHub {
     owner = "OpenVPN";
     repo = "easy-rsa";
-    rev = "v${version}";
-    sha256 = "05q60s343ydh9j6hzj0840qdcq8fkyz06q68yw4pqgqg4w68rbgs";
+    rev = gitRev;
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/easyrsa/default.nix:13]:
```suggestion
    rev = "v${version}";
```

File: pkgs/tools/networking/easyrsa/default.nix [hunk 4]
```diff
@@ -1,33 +1,42 @@
...
   installPhase = ''
-    mkdir -p $out/share/easyrsa
-    cp -r easyrsa3/{*.cnf,x509-types,vars.example} $out/share/easyrsa
-    cp easyrsa3/openssl-easyrsa.cnf $out/share/easyrsa/safessl-easyrsa.cnf
+    mkdir -p $out/share/easy-rsa
+    cp -r easyrsa3/{*.cnf,x509-types,vars.example} $out/share/easy-rsa
     install -D -m755 easyrsa3/easyrsa $out/bin/easyrsa
+
+    # Replace /usr/ with $out/ as this is where easyrsa finds its files.
+    # Also set things like the git version and date to something constant.
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/easyrsa/default.nix:25]:
```suggestion
```

File: pkgs/tools/networking/easyrsa/default.nix [hunk 5]
```diff
@@ -1,33 +1,42 @@
...
     substituteInPlace $out/bin/easyrsa \
-      --subst-var out \
-      --subst-var-by openssl ${openssl.bin}/bin/openssl
+      --replace /usr/ $out/ \
+      --replace '~VER~' '${version}' \
+      --replace '~GITHEAD~' '${gitRev}' \
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/easyrsa/default.nix:29]:
```suggestion
      --replace '~GITHEAD~' 'v${version}' \
```

File: pkgs/tools/networking/easyrsa/2.x.nix [hunk 6]
```diff
@@ -1,47 +0,0 @@
...
-, fetchFromGitHub
-, autoreconfHook
-, makeWrapper
-, gnugrep
-, openssl
-}:
-
-stdenv.mkDerivation rec {
-  pname = "easyrsa";
-  version = "2.2.0";
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/easyrsa/2.x.nix:12]: This should be an extra commit

File: pkgs/tools/networking/easyrsa/default.nix [hunk 7]
```diff
@@ -1,33 +1,37 @@
...
+    cp -r easyrsa3/{*.cnf,x509-types,vars.example} $out/share/easy-rsa
     install -D -m755 easyrsa3/easyrsa $out/bin/easyrsa
+
     substituteInPlace $out/bin/easyrsa \
-      --subst-var out \
-      --subst-var-by openssl ${openssl.bin}/bin/openssl
+      --replace /usr/ $out/ \
+      --replace '~VER~' '${version}' \
+      --replace '~GITHEAD~' 'v${version}' \
+      --replace '~DATE~' '1970-01-01'
```
Comment (numinit) [inline pkgs/tools/networking/easyrsa/default.nix:25]: I see this pattern a lot across nixpkgs and wonder if a utility function would be useful that turns the current nixpkgs version into a dummy date. 
(For example, 22.11 -> 2022-11-01).

Of course, the downside is that this forces a rebuild when nixpkgs changes, but the date is more "accurate".

May do this in a separate commit because it's come up in a handful of packages I maintain.

Comment (SuperSandro2000) [reply pkgs/tools/networking/easyrsa/default.nix:25]: > (For example, 22.11 -> 2022-11-01).

That date is for reproducibility. It would probably trigger some test failures somewhere when we change that frequently. It is better to just stick to it. Please keep it as is. For zip 1.1.1980 is fine, too.
